cpp/ql/src/semmle/code/cpp/dataflow/internal/DataFlowImpl.qll

```diff
@@ -1226,27 +1238,52 @@ private class AccessPathNil extends AccessPath, TNil {
   override AccessPathFront getFront() {
     exists(DataFlowType t | this = TNil(t) | result = TFrontNil(t))
   }
+
+  override predicate pop(Content head, AccessPath tail) { none() }
 }
 
-private class AccessPathCons extends AccessPath, TCons {
+abstract private class AccessPathCons extends AccessPath { }
+
+private class AccessPathConsNil extends AccessPathCons, TConsNil {
   override string toString() {
-    exists(Content f, int len | this = TCons(f, len) |
-      result = f.toString() + ", ... (" + len.toString() + ")"
+    exists(Content f, DataFlowType t | this = TConsNil(f, t) |
+      result = f.toString() + " : " + ppReprType(t)
     )
   }
 
   override AccessPathFront getFront() {
-    exists(Content f | this = TCons(f, _) | result = TFrontHead(f))
+    exists(Content f | this = TConsNil(f, _) | result = TFrontHead(f))
+  }
+
+  override predicate pop(Content head, AccessPath tail) {
+    exists(DataFlowType t | this = TConsNil(head, t) and tail = TNil(t))
   }
 }
 
-/** Holds if `ap0` corresponds to the cons of `f` and `ap`. */
-private predicate pop(AccessPath ap0, Content f, AccessPath ap) {
-  ap0.getFront().headUsesContent(f) and
-  consCand(f, ap.getFront(), _) and
-  ap0.len() = 1 + ap.len()
+private class AccessPathConsCons extends AccessPathCons, TConsCons {
+  override string toString() {
+    exists(Content f1, Content f2, int len | this = TConsCons(f1, f2, len) |
+      result = f1.toString() + ", " + f2.toString() + ", ... (" + len.toString() + ")"
```

Comment:
I think we should only print the length suffix when `len > 2`.

Reply:
Done.